gobblin-metrics/src/main/java/gobblin/metrics/hadoop/HadoopCounterReporter.java

```diff
@@ -0,0 +1,105 @@
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.metrics.hadoop;
+
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.hadoop.mapred.Reporter;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricFilter;
+import com.codahale.metrics.Timer;
+
+import gobblin.metrics.ContextAwareScheduledReporter;
+import gobblin.metrics.MetricContext;
+
+
+/**
+ * An implementation of {@link gobblin.metrics.ContextAwareScheduledReporter} that reports
+ * applicable metrics as Hadoop counters using a {@link org.apache.hadoop.mapred.Reporter}.
+ *
+ * @author ynli
+ */
+public class HadoopCounterReporter extends ContextAwareScheduledReporter {
+
+  private final Reporter reporter;
+
+  protected HadoopCounterReporter(MetricContext context, String name, MetricFilter filter,
+      TimeUnit rateUnit, TimeUnit durationUnit, Reporter reporter) {
+    super(context, name, filter, rateUnit, durationUnit);
+    this.reporter = reporter;
+  }
+
+  @Override
+  protected void reportInContext(MetricContext context,
+                                 SortedMap<String, Gauge> gauges,
+                                 SortedMap<String, Counter> counters,
+                                 SortedMap<String, Histogram> histograms,
+                                 SortedMap<String, Meter> meters,
+                                 SortedMap<String, Timer> timers) {
+    // Only counters are appropriate to be reported as Hadoop counters
+    reportCounters(context, counters);
+  }
+
+  /**
+   * Create a new {@link gobblin.metrics.hadoop.HadoopCounterReporter.Builder} that
+   * uses the simple name of {@link HadoopCounterReporter} as the reporter name.
+   *
+   * @param reporter a {@link org.apache.hadoop.mapred.Reporter} used to access Hadoop counters
+   * @return a new {@link gobblin.metrics.hadoop.HadoopCounterReporter.Builder}
+   */
+  public static Builder builder(Reporter reporter) {
+    return builder(HadoopCounterReporter.class.getName(), reporter);
+  }
+
+  /**
+   * Create a new {@link gobblin.metrics.hadoop.HadoopCounterReporter.Builder} that
+   * uses a given reporter name.
+   *
+   * @param name the given reporter name
+   * @param reporter a {@link org.apache.hadoop.mapred.Reporter} used to access Hadoop counters
+   * @return a new {@link gobblin.metrics.hadoop.HadoopCounterReporter.Builder}
+   */
+  public static Builder builder(String name, Reporter reporter) {
+    return new Builder(name, reporter);
+  }
+
+  private void reportCounters(MetricContext context, SortedMap<String, Counter> counters) {
+    for (Map.Entry<String, Counter> entry : counters.entrySet()) {
+      this.reporter.getCounter(context.getName(), entry.getKey()).setValue(entry.getValue().getCount());
```

Comment:
If HadoopCounterReporter is run on a mapper, reducer, etc. task, what effect will it have on the job level counters? I don't know exactly how counters are implemented.